### PR TITLE
Step 14 for the forecasting plugin

### PR DESCRIPTION
**Pull request by pritamd47**
Added [step 14](https://rat-satellitedams.readthedocs.io/en/latest/Model/ComputationalModel/#step-14) for the forecasting plugin. The `forecast` function now also converts the forecasted results to the final format.

### CODE REVIEW

**Review by SarathUW — APPROVED** [review]
Looking good to me

File: src/rat/plugins/forecasting.py [hunk 1]
```diff
@@ -379,6 +744,128 @@ def generate_forecast_state_and_inputs(
...
+            df = pd.read_csv(inflow_path, parse_dates=['date'])
+            df['inflow (m3/d)'] = df['streamflow'] * (24*60*60)        # indicate units, convert from m3/s to m3/d
+            df = df[['date', 'inflow (m3/d)']]
+
+            print(f"Converting [Inflow]: {res_name}")
+            df = df[df['date'] > basedate]
+            df.to_csv(savepath, index=False)
+            print(df.tail())
+        else:
+            print(f"Currently not displayed in website: {res_name}")
```
Comment (SanchitMinocha) [inline src/rat/plugins/forecasting.py:770]: We should not use website. Rather may be something like: Skipping {res_name} as its inflow file is not available.

File: src/rat/plugins/forecasting.py [hunk 2]
```diff
@@ -341,6 +339,373 @@ def get_GFS_data(basedate, lead_time, basin_bounds, gfs_dir):
...
+        res_scenarios = scenarios.copy()
+        forecast_inflow_fp = basin_data_dir / 'rat_outputs' / 'forecast_inflow' / f'{basedate:%Y%m%d}' / f'{res_name}.csv'
+        forecast_evap_fp = basin_data_dir / 'rat_outputs' / 'forecast_evaporation' / f'{basedate:%Y%m%d}' / f'{res_name}.csv'
+        sarea_fp = basin_data_dir / 'gee' / 'gee_sarea_tmsos' / f'{res_name}.csv'
+        aec_fp = basin_data_dir / 'final_outputs' / 'aec' / f'{res_name}.csv'
+        output_fp = basin_data_dir / 'rat_outputs' / 'forecast_outflow' / f'{basedate:%Y%m%d}' / f'{res_name}.csv'
+        output_fp.parent.mkdir(parents=True, exist_ok=True)
+
+        s_max = reservoirs[reservoirs[reservoir_shpfile_column_dict['unique_identifier']] == res_name]['CAP_MCM'].values[0]
+        grand_id = reservoirs[reservoirs[reservoir_shpfile_column_dict['unique_identifier']] == res_name]['GRAND_ID'].values[0]
```
Comment (SanchitMinocha) [inline src/rat/plugins/forecasting.py:682]: It seems to be specific for GRAND. As RAT is more generic, can we take the column name of capacity from config? 'ID' column name is already extracted from config.

File: src/rat/plugins/forecasting.py [hunk 3]
```diff
@@ -341,6 +339,373 @@ def get_GFS_data(basedate, lead_time, basin_bounds, gfs_dir):
...
+    s_max: float
+        Maximum storage capacity of the reservoir in Million cubic meters (MCM)
+    rule_curve: str
+        Path to rule curve file in .csv with headers (Month, S/Smax)   
+    st_percSmax: list
+        A list of values of storages as percentage of Smax
+    cust_st: float
+        User defined custom storage value in Million cubic meters (MCM). Used if dels_scenario = ST or None  
+    output_path: str
+        If provided, generates a netcdf file at the specified path containing the forecasted outflow results.
```
Comment (SanchitMinocha) [inline src/rat/plugins/forecasting.py:596]: There is a mismatch between docstring and the parameters that the function takes. For example, I think rat_folder has later been changed to eval_fp, area_fp and so on.

File: src/rat/plugins/forecasting.py [hunk 4]
```diff
@@ -419,8 +911,21 @@ def forecast(config, rat_logger):
...
     extracted_gfs_dir = gfs_dir / 'extracted'
     processed_gfs_dir = gfs_dir / 'processed'
-
-    for d in [raw_gefs_chirps_dir, processed_gefs_chirps_dir, raw_gfs_dir, extracted_gfs_dir, processed_gfs_dir]:
+    inflow_dst_dir = basin_data_dir / 'rat_outputs' / 'forecast_inflow' / f"{basedate:%Y%m%d}"
+    basin_reservoir_shpfile_path = Path(basin_data_dir) / 'gee' / 'gee_basin_params' / 'basin_reservoirs.shp'
+    final_inflow_out_dir = basin_data_dir / 'final_outputs' / 'forecast_inflow' / f"{basedate:%Y%m%d}"
+    final_evap_out_dir = basin_data_dir / 'final_outputs' / 'forecast_evaporation' / f"{basedate:%Y%m%d}"
+    evap_dir = basin_data_dir / 'rat_outputs' / 'forecast_evaporation' / f"{basedate:%Y%m%d}"
+    rule_curve_dir = Path('/cheetah2/pdas47/rat3_mekong/global_data/rc')    # TODO: change to dynamic
```
Comment (SanchitMinocha) [inline src/rat/plugins/forecasting.py:919]: This path needs to be changed to a generic one.

File: src/rat/plugins/forecasting.py [hunk 5]
```diff
@@ -379,6 +744,128 @@ def generate_forecast_state_and_inputs(
...
+
+            print(f"Converting [Inflow]: {res_name}")
+            df = df[df['date'] > basedate]
+            df.to_csv(savepath, index=False)
+            print(df.tail())
+        else:
+            print(f"Currently not displayed in website: {res_name}")
+
+
+def convert_forecast_evaporation(evap_dir, evap_web_dir):
```
Comment (SanchitMinocha) [inline src/rat/plugins/forecasting.py:773]: I think the parameter name should be final_evap_dir rather than eval_web_dir.

**Review by SanchitMinocha — CHANGES_REQUESTED** [review]
Haven't looked into the different forecasting_scenario functions in detail, but otherwise looks good. Left a few minor comments that needs fixing. Also, I will recommend updating the forecasting documentation as well regarding the different scenarios.

File: src/rat/plugins/forecasting.py (hunk 3 above)
Comment (pritamd47) [reply src/rat/plugins/forecasting.py:596]: Thanks for catching that, I have updated the docstring.

File: src/rat/plugins/forecasting.py (hunk 4 above)
Comment (pritamd47) [reply src/rat/plugins/forecasting.py:919]: Changed the code to read the path from the config file. The added option is under `['PLUGINS']['forecast_rule_curve_dir']`.

File: src/rat/plugins/forecasting.py (hunk 2 above)
Comment (pritamd47) [reply src/rat/plugins/forecasting.py:682]: Thanks for the suggestion. The reservoir ID and capacity of the reservoir will now be picked from the reservoir shapefile based on the `column_id` and `column_capacity` options specified in the config file.

**Review by SanchitMinocha — APPROVED** [review]
Looks good. Thank you for updating it.

**Pull request merged**